core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,839 @@
+      chunks ::= chunk
+      i += 1
+      total += chunk.length
+    }
+    chunks = chunks.reverse
+
+    if (B0 == null) Chunk.empty
+    else {
+      implicit val B = B0
+
+      val dest: Array[B] = Array.ofDim(total)
+
+      val it = chunks.iterator
+      var n  = 0
+      while (it.hasNext) {
+        val chunk = it.next
+
+        chunk.toArray(n, dest)
+        n += chunk.length
+      }
+
+      Chunk.Arr(dest)
+    }
+  }
+
+  final def foldMLazy[E, S](z: IO[E, S])(pred: S => Boolean)(f: (S, A) => IO[E, S]): IO[E, S] = {
+    val len = length
+
+    def loop(s: S, i: Int): IO[E, S] =
+      if (i >= len) IO.now(s)
+      else {
+        if (pred(s)) f(s, self(i)).flatMap(loop(_, i + 1))
+        else IO.now(s)
+      }
+
+    z.flatMap(loop(_, 0))
+  }
+
+  final def foldLeftLazy[S](z: S)(pred: S => Boolean)(f: (S, A) => S): S = {
+    val len = length
+    var s   = z
+    var i   = 0
+    while (i < len && pred(s)) {
+      s = f(s, self(i))
+      i += 1
+    }
+    s
+  }
+
+  /**
+   * Folds over the elements in this chunk from the left.
+   */
+  def foldLeft[@specialized S](s0: S)(f: (S, A) => S): S = {
+    val len = self.length
+    var s   = s0
+
+    var i = 0
+    while (i < len) {
+      s = f(s, self(i))
+      i += 1
+    }
+
+    s
+  }
+
+  /**
+   * Effectfully folds over the elements in this chunk from the left.
+   */
+  final def foldM[E, S](s: S)(f: (S, A) => IO[E, S]): IO[E, S] =
+    foldLeft[IO[E, S]](IO.now(s)) { (s, a) =>
+      s.flatMap(f(_, a))
+    }
+
+  /**
+   * Folds over the elements in this chunk from the right.
+   */
+  def foldRight[@specialized S](s0: S)(f: (A, S) => S): S = {
+    val len = self.length
+    var s   = s0
+
+    var i = len - 1
+    while (i >= 0) {
+      s = f(self(i), s)
+      i -= 1
+    }
+
+    s
+  }
+
+  override final def hashCode: Int = toArray.toSeq.hashCode
+
+  /**
+   * Returns a chunk with the elements mapped by the specified function.
+   */
+  def map[@specialized B](f: A => B): Chunk[B] = {
+    val len  = self.length
+    var dest = null.asInstanceOf[Array[B]]
+
+    var i = 0
+    while (i < len) {
+      val b = f(self(i))
+
+      if (dest == null) {
+        implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+
+        dest = Array.ofDim[B](len)
+      }
+
+      dest(i) = b
+
+      i = i + 1
+    }
+
+    if (dest != null) Chunk.Arr(dest)
+    else Chunk.Empty
+  }
+
+  /**
+   * Materializes a chunk into a chunk backed by an array. This method can
+   * improve the performance of bulk operations.
+   */
+  def materialize[A1 >: A]: Chunk[A1] = Chunk.Arr(toArray)
+
+  /**
+   * Generates a readable string representation of this chunk using the
+   * specified start, separator, and end strings.
+   */
+  final def mkString(start: String, sep: String, end: String): String = {
+    val builder = new scala.collection.mutable.StringBuilder()
+
+    builder.append(start)
+
+    var i = 0
+    while (i < self.length) {
```

Comment:
`val l = self.length`
` while (i < l) ...`

Reply:
Excellent point. Anytime we are referencing `length` or other fields, we need to copy them into the stack to avoid the overhead of `invokevirtual` on every reference.